### ISSUE TITLE
ES3 vs ES6 compilation

### ISSUE THREAD

**Issue by BobFrGit** [Microsoft/TypeScript#10465]
**TypeScript Version:**  1.8.86
**Code**

```ts
// A *self-contained* demonstration of the problem follows...

port function htmlEscape(literals, ...placeholders) {
    try {
        var result = "";

        // interleave the literals with the placeholders
        for (var i = 0; i < placeholders.length; i++) {
            result += literals[i];
            result += placeholders[i]
                .replace(/&/g, '&amp;')
                .replace(/"/g, '&quot;')
                .replace(/'/g, '&#39;')
                .replace(/</g, '&lt;')
                .replace(/>/g, '&gt;');
        }
        console.log(1);
        
        result += literals[literals.length - 1];
        return result;
    }
    catch (e) {
        debugger;
        throw e;
    }
}

```

I changed from targeting ES6 to ES3. And when I do a full rebuild it property translates in the ES3 version. But if I just change that one file and get incremental compilation I get the ES6 version.

I tried I simple repro and it always did ES3 so I"m not sure what is doing on but it's annoying.

**Comment by basarat** [comment]
Question: Are you compiling with `tsc -w`? 💟 :

**Comment by BobFrGit** [comment]
I"m using VS 2015.3.

**Comment by DanielRosenwasser** [comment]
Sounds like compile-on-save.

**Comment by BobFrGit** [comment]
Indeed compile on save -- seems like CoS isn't honoring the ES3 parameter.

**Comment by BobFrGit** [comment]
I don't know if this is related and not sure how to repro it. Often if I don't do a complete rebuild I get the following error. After a rebuild it works.

**Before the rebuild the generated codes** 
"use strict";
const http = require("http");

**After full rebuild**
"use strict";
var http = require("http");

**Source is**
import http = require("http");

OK, so I'm trying to figure this out and tried setting ES6 and it then generates the const but and gets the error even on a complete rebuild. So something else is going on -- as if generating for ES3 gave me an ES3 engine but going back to ES6 left me with an ES3 engine. I"m attaching the NSProj file 
[xp3starter.txt](https://github.com/Microsoft/TypeScript/files/430572/xp3starter.txt) to help track thsi down.



**The error message:**

Debugger listening on port 5858
Y:\dev\projects\SIte\xp3starter\xp3starter\app.js:2
const http = require("http");
^^^^^
SyntaxError: Use of const in strict mode.
    at exports.runInThisContext (vm.js:73:16)
    at Module._compile (module.js:443:25)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.runMain [as _onTimeout] (module.js:501:10)
    at Timer.listOnTimeout (timers.js:110:15)
Press any key to continue...

**Comment by BobFrGit** [comment]
I'm not sure what else I can provide beyond the project file. But I gave up the ES3/ES6 part and figured out how to install the ES6 version of Node.

**Comment by mhegazy** [comment]
@paulvanbrenk anythoughts?

**Comment by paulvanbrenk** [comment]
Can you share the project file and the `tsconfig.json` if you have one?

**Comment by mhegazy** [comment]
the project file is at: [xp3starter.txt](https://github.com/Microsoft/TypeScript/files/430572/xp3starter.txt)

**Comment by BobFrGit** [comment]
Yes, I had to rename it as txt because the uploader insisted. I'm now using ES5 because when I had it set to ES6 even rebuilds didn't get past the const problem. When I take the same program and run it with node from the command line it works fine. Perhaps there is state left over from the previous run the way VS uses the JavaScript engine?

**Comment by BobFrGit** [comment]
Note TSCONFIG j=- just package.jos

**How do I mark code so it's in a fixed with font and preserves spaces?**

{
  "name": "xp3starter",
  "version": "0.0.0",
  "description": "xp3starter",
  "main": "app.js",
  "author": {
    "name": "Bob"
  },
  "dependencies": {
    "retyped-xmldom-tsd-ambient": "^0.1.16-0",
    "xmldom": "^0.1.22"
  }
}

**Comment by basarat** [comment]
Triple back ticks (I'm using single quotes to demonstrate🌹)

'''ts 
Code
'''

**Comment by paulvanbrenk** [comment]
@BobFrGit fixed your comment ;)

**Comment by paulvanbrenk** [comment]
I think there is some caching going on in your project, since I don't repro... which version of VS are you using? And does it work after opening/closing VS?

**Comment by paulvanbrenk** [comment]
Another thing I noticed in your project file is that the options are only set in the `Debug` configuration. Can you try removing the `Condition` from this line above the TypeScript options?
```xml
<PropertyGroup Condition=" '$(Configuration)' == 'Debug' ">
```

**Comment by BobFrGit** [comment]
I only use Debug mode but I'll try removing the statement. Restarting doesn't seem to solve it. BTW, I find that sometimes I run into a problem with VS busy looking for debuging data and have to terminate all devenv's. I can get you more info next time.

**Comment by BobFrGit** [comment]
The build thing happened again. Not surprised since I was running debug all along.

**Comment by paulvanbrenk** [comment]
Can you turn on 'virtual projects' for files that are part of a solution in tools - options - text editor - typescript -projects, restart VS, open a ts file that causes issues, and make a screenshot of the Solution Explorer?

**Comment by BobFrGit** [comment]
Catching up (also trying to debug the app which is a distraction). VS is latest update 3 and all that to the best of my knowledge. Just had to blow away VS again because of the data tip busy bug 
![temp](https://cloud.githubusercontent.com/assets/5490484/17875193/59c9a070-689e-11e6-965d-a10b5c92f43e.jpg)

As to virtual projects -- it's checked for "display VP when no solutions loaded. Will check it for both cases. It says automatically compile files not part of a project. It's checked as an FYI. Restarted and found the StartUp project  had changed (I have a client HTML project though not using it yet).

No to wait till the problem reappears.

**Comment by BobFrGit** [comment]
OK, just failed again

'```Debugger listening on port 5858
'```Y:\dev\projects\SIte\xp3starter\xp3starter\app.js:2
'```const http = require("http");
'```^^^^^
```SyntaxError: Use of const in strict mode.
```    at exports.runInThisContext (vm.js:73:16)
```    at Module._compile (module.js:443:25)
```    at Object.Module._extensions..js (module.js:478:10)
```    at Module.load (module.js:355:32)
```    at Function.Module._load (module.js:310:12)
```    at Module.runMain [as _onTimeout] (module.js:501:10)
```    at Timer.listOnTimeout (timers.js:110:15)
```Press any key to continue...

(Three back ticks per line? 🙁)

**Comment by paulvanbrenk** [comment]
Is the file included in the virtual project, which should show in the Solution Explorer?

**Comment by BobFrGit** [comment]
One hint is that the old nod session stay around. If you look at this pile you'll see a number of failures. Shouldn't terminating a session close the console window?
![temp2](https://cloud.githubusercontent.com/assets/5490484/17875781/fe04e448-68a2-11e6-9ec4-ea3e625cb53b.jpg)

**Comment by paulvanbrenk** [comment]
@mousetraps any insights here about the node session hanging around after debugging?

**Comment by BobFrGit** [comment]
Solution explorer not sure if this useful but in response to the Solution Explorer question. (wonder why some images show up and others don't.)

![temp3](https://cloud.githubusercontent.com/assets/5490484/17876022/f3a0685e-68a4-11e6-88b1-a31b27c41b88.jpg)

**Comment by BobFrGit** [comment]
Just as an FYI

```npm WARN You are using a pre-release version of node and things may not work as expected

```y:\tmpx\tx>node -v
```v5.7.1-0

I don't think it should make a difference but maybe ...

**Comment by mousetraps** [comment]
Regarding the Node console hanging around after the debug session has ended - that is by design in order to handle cases where the session has terminated, but you still want some time to read the console output (rather than printing stuff and closing before you got the chance to read it). Unfortunately this results in a barage of console windows unless you explicitly close each one via press any key or simply closing the window. To disable this behavior, you can go to the Node.js Tools settings in the options dialog and uncheck the "wait for input when process exits" optionsAnother idea that wed be open to as a contribution is closing the previous console window when starting a new session so there would only be one open at any given time. Sent from my Rotary Phone


		_____________________________

**Comment by BobFrGit** [comment]
I tried to repro it and noticed the console session went away by itself so it only seems to happen on abnormal termination. It also inspired me to change the default size of the windows. Sent by tapping my switch-hook.

**Comment by paulvanbrenk** [comment]
I tried this with several projects in TypeScript 2.0 Beta and don't reproduce this...

Thanks for reporting, but closing this as fixed in TypeScript 2.0.
